### PR TITLE
Move gossipsub into a separate crate

### PR DESCRIPTION
**Pull request by jxs**
## Issue Addressed

this will helps us better track changes into the Gossipsub protocol into different versions and a proper `CHANGELOG.md` file.
Eventually we can also release it as a separate crate.

This PR also probably addresses https://github.com/sigp/lighthouse/issues/5389 cc @pawanjay176

### CODE REVIEW

**Review by AgeManning — COMMENTED** [review]
The changes look good. 

I think (if it's possible) it might be better to put it inside the lighthouse_network dir as it seems to be the best place logically for it. Avoid the consensus people getting too scared with seeing it inside beacon_node :sweat_smile:

**Comment by jxs** [comment]
> The changes look good.
> 
> I think (if it's possible) it might be better to put it inside the lighthouse_network dir as it seems to be the best place logically for it. Avoid the consensus people getting too scared with seeing it inside beacon_node 😅

yeah I thought of that, but unfortunately it's not possible as `lighthouse_network` is a crate by itself

**Comment by AgeManning** [comment]
Oh. Isn't `beacon_node` itself also a crate?

Or another example is the `validator_client` :https://github.com/sigp/lighthouse/blob/stable/validator_client/Cargo.toml
which also has `slashing_protection` inside its directory. 

These looked similar to `lighthouse_network`, does cargo throw an error if we try?

**Review by jxs — COMMENTED** [review]
>Oh. Isn't beacon_node itself also a crate?
>
>Or another example is the validator_client >:https://github.com/sigp/lighthouse/blob/stable/validator_client/Cargo.toml
>which also has slashing_protection inside its directory.
>
>These looked similar to lighthouse_network, does cargo throw an error if we try?

Right, thanks for pushing this age, it's actually possible. I naively assumed that it wasn't as I didn't see any other example on the workspace, but it's perfectly fine. Updated!

**Comment by AgeManning** [comment]
Oh nice. I think this structure is better. 

Should we put the CHANGELOG inside gossipsub tho?

File: beacon_node/lighthouse_network/gossipsub/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+bytes = "1.5"
+either = "1.9"
+fnv = "1.0.7"
+futures = "0.3.30"
+futures-ticker = "0.0.3"
+futures-timer = "3.0.2"
+getrandom = "0.2.12"
+hex_fmt = "0.3.0"
+instant = "0.1.12"
+libp2p = { version = "0.53", default-features = false, features = ["identify", "yamux", "noise", "dns", "tcp", "tokio", "plaintext", "secp256k1", "macros", "ecdsa", "metrics", "quic", "upnp"]}
```
Comment (ackintosh) [inline beacon_node/lighthouse_network/gossipsub/Cargo.toml]: I think the `features` section can be removed.

```suggestion
libp2p = { version = "0.53", default-features = false }
```

Comment (jxs) [reply beacon_node/lighthouse_network/gossipsub/Cargo.toml]: thanks Akihito, updated!

**Comment by ackintosh** [comment]
Regarding to logging, we need to change the target name from `libp2p_gossipsub` to `gossipsub` so that the gossipsub logs are correctly written to a file. This change will resolve #5389. 💡 

https://github.com/sigp/lighthouse/blob/612eaf2d41b9ca98d7e47b327f4564d41ff24ac9/common/logging/src/tracing_logging_layer.rs#L29-L33

**Review by AgeManning — APPROVED** [review]

**Comment by AgeManning** [comment]
@mergifyio queue

**Pull request merged**